File: src/toubkal/app/core/diagnostics/error-tracker.test.ts
/**
 * Error Tracker Tests
 * 
 * Comprehensive test suite for the Error Tracker system.
 */

import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { ErrorTracker, ErrorSeverity, ErrorCategory } from './error-tracker';

// Mock the Logger
const mockLogger = {
  info: vi.fn(),
  warn: vi.fn(),
  error: vi.fn(),
  debug: vi.fn()
};

vi.mock('./logger', () => ({
  logger: {
    getInstance: vi.fn(() => mockLogger)
  }
}));

describe('ErrorTracker', () => {
  let errorTracker: ErrorTracker;

  beforeEach(() => {
    errorTracker = ErrorTracker.getInstance();
    
    // Clear any existing state
    errorTracker.clearErrors();
    
    // Clear mock calls
    mockLogger.info.mockClear();
    mockLogger.warn.mockClear();
    mockLogger.error.mockClear();
    mockLogger.debug.mockClear();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  describe('initialization', () => {
    it('should initialize with default config', () => {
      errorTracker.initialize();
      
      expect(mockLogger.info).toHaveBeenCalledWith(
        'ErrorTracker',
        'Error tracking system initialized',
        expect.objectContaining({
          maxReports: 1000,
          privacyMode: true
        })
      );
    });

    it('should initialize with custom config', () => {
      const customConfig = {
        maxReports: 500,
        reportRetentionDays: 7,
        enableAutoReporting: false,
        privacyMode: false
      };
      
      errorTracker.initialize(customConfig);
      
      expect(mockLogger.info).toHaveBeenCalledWith(
        'ErrorTracker',
        'Error tracking system initialized',
        expect.objectContaining({
          maxReports: 500,
          privacyMode: false
        })
      );
    });
  });

  describe('error tracking', () => {
    beforeEach(() => {
      errorTracker.initialize();
    });

    it('should track a new error', () => {
      const error = new Error('Test error message');
      const errorId = errorTracker.trackError(
        error,
        ErrorSeverity.HIGH,
        ErrorCategory.System,
        { component: 'test-component', action: 'test-action' }
      );

      expect(errorId).toBeDefined();
      expect(mockLogger.error).toHaveBeenCalledWith(
        'ErrorTracker',
        'New error tracked',
        expect.objectContaining({
          severity: ErrorSeverity.HIGH,
          category: ErrorCategory.System,
          message: 'Test error message'
        })
      );
    });

    it('should track the same error multiple times', () => {
      const error = new Error('Duplicate error');
      const context = { component: 'test', action: 'test' };
      
      const errorId1 = errorTracker.trackError(error, ErrorSeverity.MEDIUM, ErrorCategory.UNKNOWN, context);
      const errorId2 = errorTracker.trackError(error, ErrorSeverity.MEDIUM, ErrorCategory.UNKNOWN, context);

      expect(errorId1).toBe(errorId2);
      
      const report = errorId1;
      expect(report).toBeDefined();
      expect(report?.count).toBe(2);
    });

    it('should generate unique IDs for different errors', () => {
      const error1 = new Error('Error 1');
      const error2 = new Error('Error 2');
      
      const id1 = errorTracker.trackError(error1);
      const id2 = errorTracker.trackError(error2);

      expect(id1).not.toBe(id2);
    });

    it('should update severity and category for existing errors', () => {
      const error = new Error('Test error');
      const context = { component: 'test', action: 'test' };
      
      const errorId = errorTracker.trackError(error, ErrorSeverity.LOW, ErrorCategory.UNKNOWN, context);
      errorTracker.trackError(error, ErrorSeverity.HIGH, ErrorCategory.SECURITY, context);
      
      const report = errorId;
      expect(report?.severity).toBe(ErrorSeverity.HIGH);
      expect(report?.category).toBe(ErrorCategory.SECURITY);
      expect(report?.count).toBe(2);
    });
  });

  describe('error retrieval', () => {
    beforeEach(() => {
      errorTracker.initialize();
    });

    it('should get error report by ID', () => {
      const error = new Error('Test error');
      const errorId = errorTracker.trackError(error);
      
      const report = errorId;
      expect(report).toBeDefined();
      expect(report?.message).toBe('Test error');
      expect(report?.count).toBe(1);
    });

    it('should return undefined for non-existent error ID', () => {
      const report = errorTracker.getErrorReport('non-existent-id');
      expect(report).toBeUndefined();
    });

    it('should get all errors', () => {
      errorTracker.trackError(new Error('Error 1'));
      errorTracker.trackError(new Error('Error 2'));
      
      const allErrors = errorTracker.getAllErrors();
      expect(allErrors).toHaveLength(2);
    });

    it('should get errors by severity', () => {
      errorTracker.trackError(new Error('High error'), ErrorSeverity.HIGH);
      errorTracker.trackError(new Error('Low error'), ErrorSeverity.LOW);
      errorTracker.trackError(new Error('Another high error'), ErrorSeverity.HIGH);
      
      const highErrors = errorTracker.getErrorsBySeverity(ErrorSeverity.HIGH);
      expect(highErrors).toHaveLength(2);
      
      const lowErrors = errorTracker.getErrorsBySeverity(ErrorSeverity.LOW);
      expect(lowErrors).toHaveLength(1);
    });

    it('should get errors by category', () => {
      errorTracker.trackError(new Error('Network error'), ErrorSeverity.MEDIUM, ErrorCategory.NETWORK);
      errorTracker.trackError(new Error('Security error'), ErrorSeverity.HIGH, ErrorCategory.SECURITY);
      errorTracker.trackError(new Error('Another network error'), ErrorSeverity.MEDIUM, ErrorCategory.NETWORK);
      
      const networkErrors = errorTracker.getErrorsByCategory(ErrorCategory.NETWORK);
      expect(networkErrors).toHaveLength(2);
      
      const securityErrors = errorTracker.getErrorsByCategory(ErrorCategory.SECURITY);
      expect(securityErrors).toHaveLength(1);
    });
  });

  describe('error statistics', () => {
    beforeEach(() => {
      errorTracker.initialize();
    });

    it('should generate error statistics', () => {
      // Create some test errors
      errorTracker.trackError(new Error('High error'), ErrorSeverity.HIGH, ErrorCategory.System);
      errorTracker.trackError(new Error('Medium error'), ErrorSeverity.MEDIUM, ErrorCategory.NETWORK);
      errorTracker.trackError(new Error('Another high error'), ErrorSeverity.HIGH, ErrorCategory.SECURITY);
      
      const stats = errorTracker.getErrorStats();
      
      expect(stats.totalErrors).toBe(3);
      expect(stats.errorsBySeverity[ErrorSeverity.HIGH]).toBe(2);
      expect(stats.errorsBySeverity[ErrorSeverity.MEDIUM]).toBe(1);
      expect(stats.errorsByCategory[ErrorCategory.System]).toBe(1);
      expect(stats.errorsByCategory[ErrorCategory.NETWORK]).toBe(1);
      expect(stats.errorsByCategory[ErrorCategory.SECURITY]).toBe(1);
      expect(stats.topErrors).toHaveLength(3);
      expect(stats.recentErrors).toHaveLength(3);
    });

    it('should calculate error rate correctly', () => {
      errorTracker.trackError(new Error('Test error'));
      
      const stats = errorTracker.getErrorStats(1); // 1 hour window
      expect(stats.errorRate).toBeGreaterThan(0);
    });
  });

  describe('error resolution', () => {
    beforeEach(() => {
      errorTracker.initialize();
    });

    it('should mark error as resolved', () => {
      const error = new Error('Test error');
      const errorId = errorTracker.trackError(error);
      
      const result = errorTracker.markErrorResolved(errorId.id);
      expect(result).toBe(true);
      
      const report = errorId;
      expect(report?.resolved).toBe(true);
    });

    it('should return false for non-existent error ID', () => {
      const result = errorTracker.markErrorResolved('non-existent-id');
      expect(result).toBe(false);
    });
  });

  describe('error reporting', () => {
    beforeEach(() => {
      errorTracker.initialize();
    });

    it('should report error', () => {
      const error = new Error('Test error');
      const errorId = errorTracker.trackError(error);
      
      const result = errorTracker.reportError(errorId.id);
      expect(result).toBe(true);
      
      expect(mockLogger.error).toHaveBeenCalledWith(
        'ErrorTracker',
        'Error report generated',
        expect.objectContaining({
          errorId,
          message: 'Test error'
        })
      );
    });

    it('should return false for non-existent error ID', () => {
      const result = errorTracker.reportError('non-existent-id');
      expect(result).toBe(false);
    });

    it('should auto-report critical errors when enabled', () => {
      // Ensure auto-reporting is enabled
      errorTracker.initialize({
        enableAutoReporting: true
      });
      
      const error = new Error('Critical error');
      errorTracker.trackError(error, ErrorSeverity.CRITICAL);
      
      // Should have been auto-reported - check for both the initial tracking and the report
      expect(mockLogger.error).toHaveBeenCalledWith(
        'ErrorTracker',
        'New error tracked',
        expect.any(Object)
      );
      
      // The auto-report should also be called - check that it was called at least once
      expect(mockLogger.error).toHaveBeenCalledTimes(2);
      expect(mockLogger.error).toHaveBeenCalledWith(
        'ErrorTracker',
        'Error report generated',
        expect.any(Object)
      );
    });
  });

  describe('error export and cleanup', () => {
    beforeEach(() => {
      errorTracker.initialize();
    });

    it('should export errors as JSON', () => {
      errorTracker.trackError(new Error('Test error 1'));
      errorTracker.trackError(new Error('Test error 2'));
      
      const exported = errorTracker.exportErrors();
      expect(exported).toBeDefined();
      
      const parsed = JSON.parse(exported);
      expect(parsed).toHaveLength(2);
      expect(parsed[0]).toHaveProperty('id');
      expect(parsed[0]).toHaveProperty('message');
      expect(parsed[0]).toHaveProperty('severity');
      expect(parsed[0]).toHaveProperty('category');
    });

    it('should clear all errors', () => {
      errorTracker.trackError(new Error('Test error 1'));
      errorTracker.trackError(new Error('Test error 2'));
      
      expect(errorTracker.getAllErrors()).toHaveLength(2);
      
      errorTracker.clearErrors();
      
      expect(errorTracker.getAllErrors()).toHaveLength(0);
      expect(mockLogger.info).toHaveBeenCalledWith(
        'ErrorTracker',
        'All error reports cleared'
      );
    });
  });

  describe('error context and metadata', () => {
    beforeEach(() => {
      errorTracker.initialize();
    });

    it('should preserve error context', () => {
      const error = new Error('Test error');
      const context = {
        userId: 'user123',
        sessionId: 'session456',
        url: 'https://example.com',
        userAgent: 'Mozilla/5.0',
        component: 'test-component',
        action: 'test-action',
        metadata: { key: 'value' }
      };
      
      const errorId = errorTracker.trackError(error, ErrorSeverity.MEDIUM, ErrorCategory.UNKNOWN, context);
      const report = errorId;
      
      expect(report?.context.userId).toBe('user123');
      expect(report?.context.sessionId).toBe('session456');
      expect(report?.context.url).toBe('https://example.com');
      expect(report?.context.userAgent).toBe('Mozilla/5.0');
      expect(report?.context.component).toBe('test-component');
      expect(report?.context.action).toBe('test-action');
      expect(report?.context.metadata).toEqual({ key: 'value' });
    });

    it('should generate appropriate tags', () => {
      const error = new Error('Network fetch failed');
      const errorId = errorTracker.trackError(error, ErrorSeverity.HIGH, ErrorCategory.NETWORK);
      const report = errorId;
      
      expect(report?.tags).toContain('high');
      expect(report?.tags).toContain('network');
      expect(report?.tags).toContain('network'); // Content-based tag
    });
  });

  describe('global error handlers', () => {
    beforeEach(() => {
      errorTracker.initialize();
    });

    it('should handle uncaught errors', () => {
      // Mock window and addEventListener
      const mockAddEventListener = vi.fn();
      Object.defineProperty(global, 'window', {
        value: {
          addEventListener: mockAddEventListener,
          location: { href: 'https://example.com' }
        },
        writable: true
      });
      
      Object.defineProperty(global, 'navigator', {
        value: { userAgent: 'Mozilla/5.0' },
        writable: true
      });

      // Re-initialize to set up handlers
      errorTracker.initialize();
      
      expect(mockAddEventListener).toHaveBeenCalledWith('error', expect.any(Function));
      expect(mockAddEventListener).toHaveBeenCalledWith('unhandledrejection', expect.any(Function));
    });
  });

  describe('error cleanup and retention', () => {
    it('should cleanup old reports when max reports exceeded', () => {
      const config = {
        maxReports: 2,
        reportRetentionDays: 1
      };
      
      errorTracker.initialize(config);
      
      // Add more errors than maxReports
      errorTracker.trackError(new Error('Error 1'));
      errorTracker.trackError(new Error('Error 2'));
      errorTracker.trackError(new Error('Error 3'));
      
      // Should have cleaned up to maxReports
      expect(errorTracker.getAllErrors().length).toBeLessThanOrEqual(2);
    });
  });
});

File: src/toubkal/components/diagnostics/error_tracker_test.cc
#include "toubkal/components/diagnostics/error_tracker.h"

#include <gtest/gtest.h>
#include <memory>
#include <chrono>
#include "base/test/scoped_task_environment.h"

namespace toubkal {
namespace diagnostics {

class ErrorTrackerTest : public testing::Test {
 protected:
  void SetUp() override {
    error_tracker_ = ErrorTracker::GetInstance();
    error_tracker_->ClearErrors();
    
    // Initialize with test config
    ErrorTrackerConfig config;
    config.max_reports = 100;
    config.report_retention_days = 1;
    config.enable_auto_reporting = true;
    config.enable_user_feedback = true;
    config.enable_crash_reporting = true;
    config.privacy_mode = true;
    
    error_tracker_->Initialize(config);
  }

  void TearDown() override {
    error_tracker_->ClearErrors();
  }

  ErrorTracker* error_tracker_;
};

TEST_F(ErrorTrackerTest, TrackError) {
  const std::string error_message = "Test error message";
  const std::string error_id = error_tracker_->TrackError(
      error_message, ErrorSeverity::HIGH, ErrorCategory::SYSTEM);
  
  EXPECT_FALSE(error_id.empty());
  
  ErrorReport* report = error_tracker_->GetErrorReport(error_id);
  ASSERT_NE(report, nullptr);
  EXPECT_EQ(report->message, error_message);
  EXPECT_EQ(report->severity, ErrorSeverity::HIGH);
  EXPECT_EQ(report->category, ErrorCategory::SYSTEM);
  EXPECT_EQ(report->count, 1);
  EXPECT_FALSE(report->resolved);
}

TEST_F(ErrorTrackerTest, TrackErrorWithException) {
  const std::runtime_error exception("Runtime error");
  const std::string error_id = error_tracker_->TrackError(
      exception, ErrorSeverity::CRITICAL, ErrorCategory::SYSTEM);
  
  EXPECT_FALSE(error_id.empty());
  
  ErrorReport* report = error_tracker_->GetErrorReport(error_id);
  ASSERT_NE(report, nullptr);
  EXPECT_EQ(report->message, "Runtime error");
  EXPECT_EQ(report->severity, ErrorSeverity::CRITICAL);
  EXPECT_EQ(report->category, ErrorCategory::SYSTEM);
}

TEST_F(ErrorTrackerTest, TrackErrorWithContext) {
  ErrorContext context;
  context.user_id = "user123";
  context.session_id = "session456";
  context.url = "https://example.com";
  context.user_agent = "Mozilla/5.0";
  context.component = "test-component";
  context.action = "test-action";
  context.metadata["key"] = "value";
  
  const std::string error_id = error_tracker_->TrackError(
      "Test error", ErrorSeverity::MEDIUM, ErrorCategory::NETWORK, context);
  
  ErrorReport* report = error_tracker_->GetErrorReport(error_id);
  ASSERT_NE(report, nullptr);
  EXPECT_EQ(report->context.user_id, "user123");
  EXPECT_EQ(report->context.session_id, "session456");
  EXPECT_EQ(report->context.url, "https://example.com");
  EXPECT_EQ(report->context.user_agent, "Mozilla/5.0");
  EXPECT_EQ(report->context.component, "test-component");
  EXPECT_EQ(report->context.action, "test-action");
  EXPECT_EQ(report->context.metadata["key"], "value");
}

TEST_F(ErrorTrackerTest, TrackDuplicateError) {
  const std::string error_message = "Duplicate error";
  const std::string error_id1 = error_tracker_->TrackError(
      error_message, ErrorSeverity::LOW, ErrorCategory::UNKNOWN);
  const std::string error_id2 = error_tracker_->TrackError(
      error_message, ErrorSeverity::LOW, ErrorCategory::UNKNOWN);
  
  EXPECT_EQ(error_id1, error_id2);
  
  ErrorReport* report = error_tracker_->GetErrorReport(error_id1);
  ASSERT_NE(report, nullptr);
  EXPECT_EQ(report->count, 2);
}

TEST_F(ErrorTrackerTest, GetAllErrors) {
  error_tracker_->TrackError("Error 1", ErrorSeverity::LOW);
  error_tracker_->TrackError("Error 2", ErrorSeverity::MEDIUM);
  error_tracker_->TrackError("Error 3", ErrorSeverity::HIGH);
  
  std::vector<ErrorReport> all_errors = error_tracker_->GetAllErrors();
  EXPECT_EQ(all_errors.size(), 3);
}

TEST_F(ErrorTrackerTest, GetErrorsBySeverity) {
  error_tracker_->TrackError("High Error 1", ErrorSeverity::HIGH);
  error_tracker_->TrackError("Low Error 1", ErrorSeverity::LOW);
  error_tracker_->TrackError("High Error 2", ErrorSeverity::HIGH);
  
  std::vector<ErrorReport> high_errors = error_tracker_->GetErrorsBySeverity(ErrorSeverity::HIGH);
  EXPECT_EQ(high_errors.size(), 2);
  
  std::vector<ErrorReport> low_errors = error_tracker_->GetErrorsBySeverity(ErrorSeverity::LOW);
  EXPECT_EQ(low_errors.size(), 1);
}

TEST_F(ErrorTrackerTest, GetErrorsByCategory) {
  error_tracker_->TrackError("Network Error 1", ErrorSeverity::MEDIUM, ErrorCategory::NETWORK);
  error_tracker_->TrackError("Security Error 1", ErrorSeverity::HIGH, ErrorCategory::SECURITY);
  error_tracker_->TrackError("Network Error 2", ErrorSeverity::MEDIUM, ErrorCategory::NETWORK);
  
  std::vector<ErrorReport> network_errors = error_tracker_->GetErrorsByCategory(ErrorCategory::NETWORK);
  EXPECT_EQ(network_errors.size(), 2);
  
  std::vector<ErrorReport> security_errors = error_tracker_->GetErrorsByCategory(ErrorCategory::SECURITY);
  EXPECT_EQ(security_errors.size(), 1);
}

TEST_F(ErrorTrackerTest, GetErrorStats) {
  error_tracker_->TrackError("High Error", ErrorSeverity::HIGH, ErrorCategory::SYSTEM);
  error_tracker_->TrackError("Medium Error", ErrorSeverity::MEDIUM, ErrorCategory::NETWORK);
  error_tracker_->TrackError("Another High Error", ErrorSeverity::HIGH, ErrorCategory::SECURITY);
  
  ErrorStats stats = error_tracker_->GetErrorStats();
  
  EXPECT_EQ(stats.total_errors, 3);
  EXPECT_EQ(stats.errors_by_severity[ErrorSeverity::HIGH], 2);
  EXPECT_EQ(stats.errors_by_severity[ErrorSeverity::MEDIUM], 1);
  EXPECT_EQ(stats.errors_by_category[ErrorCategory::SYSTEM], 1);
  EXPECT_EQ(stats.errors_by_category[ErrorCategory::NETWORK], 1);
  EXPECT_EQ(stats.errors_by_category[ErrorCategory::SECURITY], 1);
  EXPECT_GT(stats.error_rate, 0.0);
}

TEST_F(ErrorTrackerTest, MarkErrorResolved) {
  const std::string error_id = error_tracker_->TrackError("Test error");
  
  ErrorReport* report = error_tracker_->GetErrorReport(error_id);
  ASSERT_NE(report, nullptr);
  EXPECT_FALSE(report->resolved);
  
  bool result = error_tracker_->MarkErrorResolved(error_id);
  EXPECT_TRUE(result);
  
  report = error_tracker_->GetErrorReport(error_id);
  ASSERT_NE(report, nullptr);
  EXPECT_TRUE(report->resolved);
}

TEST_F(ErrorTrackerTest, MarkNonExistentErrorResolved) {
  bool result = error_tracker_->MarkErrorResolved("non-existent-id");
  EXPECT_FALSE(result);
}

TEST_F(ErrorTrackerTest, ReportError) {
  const std::string error_id = error_tracker_->TrackError("Test error");
  
  bool result = error_tracker_->ReportError(error_id);
  EXPECT_TRUE(result);
}

TEST_F(ErrorTrackerTest, ReportNonExistentError) {
  bool result = error_tracker_->ReportError("non-existent-id");
  EXPECT_FALSE(result);
}

TEST_F(ErrorTrackerTest, ExportErrors) {
  error_tracker_->TrackError("Error 1", ErrorSeverity::LOW);
  error_tracker_->TrackError("Error 2", ErrorSeverity::HIGH);
  
  std::string exported = error_tracker_->ExportErrors();
  EXPECT_FALSE(exported.empty());
  
  // Should be valid JSON array
  EXPECT_TRUE(exported.find("[") != std::string::npos);
  EXPECT_TRUE(exported.find("]") != std::string::npos);
}

TEST_F(ErrorTrackerTest, ClearErrors) {
  error_tracker_->TrackError("Error 1");
  error_tracker_->TrackError("Error 2");
  
  std::vector<ErrorReport> all_errors = error_tracker_->GetAllErrors();
  EXPECT_EQ(all_errors.size(), 2);
  
  error_tracker_->ClearErrors();
  
  all_errors = error_tracker_->GetAllErrors();
  EXPECT_EQ(all_errors.size(), 0);
}

TEST_F(ErrorTrackerTest, GetNonExistentErrorReport) {
  ErrorReport* report = error_tracker_->GetErrorReport("non-existent-id");
  EXPECT_EQ(report, nullptr);
}

TEST_F(ErrorTrackerTest, ErrorIdGeneration) {
  const std::string error1 = error_tracker_->TrackError("Error 1");
  const std::string error2 = error_tracker_->TrackError("Error 2");
  
  EXPECT_NE(error1, error2);
  EXPECT_FALSE(error1.empty());
  EXPECT_FALSE(error2.empty());
}

TEST_F(ErrorTrackerTest, ErrorTags) {
  const std::string error_id = error_tracker_->TrackError(
      "Network fetch failed", ErrorSeverity::HIGH, ErrorCategory::NETWORK);
  
  ErrorReport* report = error_tracker_->GetErrorReport(error_id);
  ASSERT_NE(report, nullptr);
  
  // Should have severity and category tags
  EXPECT_TRUE(std::find(report->tags.begin(), report->tags.end(), "2") != report->tags.end()); // HIGH = 2
  EXPECT_TRUE(std::find(report->tags.begin(), report->tags.end(), "0") != report->tags.end()); // NETWORK = 0
  
  // Should have content-based tags
  EXPECT_TRUE(std::find(report->tags.begin(), report->tags.end(), "network") != report->tags.end());
}

TEST_F(ErrorTrackerTest, ErrorRetention) {
  ErrorTrackerConfig config;
  config.max_reports = 2;
  config.report_retention_days = 1;
  config.enable_auto_reporting = true;
  config.enable_user_feedback = true;
  config.enable_crash_reporting = true;
  config.privacy_mode = true;
  
  error_tracker_->Initialize(config);
  
  // Add more errors than max_reports
  error_tracker_->TrackError("Error 1");
  error_tracker_->TrackError("Error 2");
  error_tracker_->TrackError("Error 3");
  
  // Should have cleaned up to max_reports
  std::vector<ErrorReport> all_errors = error_tracker_->GetAllErrors();
  EXPECT_LE(all_errors.size(), 2);
}

} // namespace diagnostics
} // namespace toubkal
